processor/src/advice/mod.rs

```diff
@@ -7,188 +7,113 @@ use vm_core::{
+/// 3. Provide advice sets, that are mappings from a Merkle root its tree. The tree should yield
+///    nodes & leaves, and will provide a Merkle path if a leaf is updated.
+pub trait AdviceProvider {
     // ADVICE TAPE
     // --------------------------------------------------------------------------------------------
 
-    /// Removes the next element from the advice tape and returns it.
+    /// Pop an element from the advice tape and returns it.
     ///
     /// # Errors
     /// Returns an error if the advice tape is empty.
-    pub fn read_tape(&mut self) -> Result<Felt, ExecutionError> {
-        self.tape.pop().ok_or(ExecutionError::AdviceTapeReadFailed(self.step))
-    }
+    fn read_tape(&mut self) -> Result<Felt, ExecutionError>;
 
-    /// Removes a word (4 elements) from the advice tape and returns it.
+    /// Pops a word (4 elements) from the advice tape and returns it.
+    ///
+    /// Note: a word is always stored as little-endian. A `[...,a,b,c,d]` tape will yield
+    /// `[d,c,b,a]`.
     ///
     /// # Errors
     /// Returns an error if the advice tape does not contain a full word.
-    pub fn read_tapew(&mut self) -> Result<Word, ExecutionError> {
-        if self.tape.len() < 4 {
-            return Err(ExecutionError::AdviceTapeReadFailed(self.step));
-        }
-
-        let idx = self.tape.len() - 4;
-        let result = [self.tape[idx + 3], self.tape[idx + 2], self.tape[idx + 1], self.tape[idx]];
-
-        self.tape.truncate(idx);
-
-        Ok(result)
-    }
+    fn read_tape_w(&mut self) -> Result<Word, ExecutionError>;
 
-    /// Removes the next two words from the advice tape and returns them.
+    /// Pops a double word (8 elements) from the advice tape and returns them.
+    ///
+    /// Note: a double word is always stored as little-endian. A `[...,a,b,c,d,e,f,g,h]` tape will
+    /// yield `[h,g,f,e,d,c,b,a]`.
     ///
     /// # Errors
     /// Returns an error if the advice tape does not contain two words.
-    pub fn read_tape_double(&mut self) -> Result<[Word; 2], ExecutionError> {
-        let word0 = self.read_tapew()?;
-        let word1 = self.read_tapew()?;
-
-        Ok([word0, word1])
-    }
+    fn read_tape_dw(&mut self) -> Result<[Word; 2], ExecutionError>;
 
     /// Writes the provided value at the head of the advice tape.
-    pub fn write_tape(&mut self, value: Felt) {
-        self.tape.push(value);
-    }
+    fn write_tape(&mut self, value: Felt);
 
-    /// Retrieves a list of elements from a key-value map for the specified key, reverses it, and
-    /// writes the reversed list at the head of the advice tape. This way, the first element in the
-    /// list is located at the head of the advice tape.
+    /// Fetch a keyed tape from the values map, reversing and appending it to the advice tape.
+    ///
+    /// Note: this operation shouldn't consume the map element so it can be called multiple times
+    /// for the same key.
```

Comment:
nit: I'd change "shouldn't" to "doesn't" here.

Reply:
This is an interesting point of discussion. I often like to use weak imperatives for traits.

Something like: a concrete implementation is `won't`, and a contract/trait is `shouldn't`.

The reason for that is the nature of a trait, that is to set guidelines for expected behavior, while these decisions falls ultimately to the implementer. We can of course verbally enforce that, but it will not reflect what happens in the code where the implementer has total control and can easily contradict what is described on the documentation.

Reply:
Ah - makes sense! I'm fine either way then.

Reply:
My two cents on this conversation.

> The reason for that is the nature of a trait, that is to set guidelines for expected behavior, while these decisions falls ultimately to the implementer.

I don't agree with this point. IMO a trait is not a guideline but a contract. The trait designer is the one that has freedom to decide what is part of the contract, once that is done implementations _must_ follow the contract.

Sometimes the trait is designed to required something. E.g. the [`From`](https://doc.rust-lang.org/std/convert/trait.From.html) trait **must not** fail. That is part of the contract, and API users can rely on that. The decision to fail does that fall to the implementer, if the implementation fails it is a bug.

We can also have situations were things are unspecified, and this is what we have here when you say `operation shouldn't consume the map element`. You made a design decision, and the design here is to give no guarantees for the caller. If the caller assumes the key is not removed and uses a loop, the code is most likely buggy, because you allowed for some implementation to remove the key from the map.

Because of that IMO we should either remove the suggestion of using it in a loop, because if one did that, i relies on unspecified behavior. Or change the wording to *must*, so that this kind of code can be written.